### ISSUE TITLE
How to check the way parameters get sharded into devices?

### ISSUE THREAD

**Issue by hyhuang00** [facebookresearch/fairscale#955]
I have checked through the documentation, but I didn't find a good function that can tell me which layer is on which device, as suggested by the [tutorial](https://fairscale.readthedocs.io/en/latest/tutorials/pipe.html). 

By the way, the tutorial refers to https://github.com/facebookresearch/fairscale/blob/main/examples/tutorial_pipe.py for more information, but it seems like this link is no longer valid.

**Comment by min-xu-ai** [comment]
@hyhuang00 thanks for raising the issue. We are phasing out support for pipe since it is already presented in pytorch upstream code. Perhaps you can give the pytorch version a try?

To get which layer is on which device, mainly it depends on how you split the model at the beginning. I remember that user specifies how to split so the user should know which layer is on which device based on the split she provided. Hope this answers your question.

**Comment by hyhuang00** [comment]
Thank you for your suggestion. I am using some heavily wrapped scripts from [fairseq](https://github.com/pytorch/fairseq) and trying to understand how the layers of a model get sharded by FSDP. It seems like the standard `torch.tensor.get_device()` will return invalid values (-1) in this case. For FSDP, it seems like the user need to provide a process group to the init function. Does that mean as long as I can identify where the process group are initialized from, parameters for the wrapped module are successfully located? I thought the FSDP wrappers may perform multi-device data parallelism, but it seems like it's not the case. Please correct me if I'm wrong since I'm a newcomer to parallelism.

**Comment by min-xu-ai** [comment]
Sorry. I am quite confused by your questions. I thought you were asking about pipeline parallelism. But you second comment seems to be asking about FSDP.

Yes, FSDP will perform multi-device data parallelism. There is a blog introducing how FSDP work.

https://engineering.fb.com/2021/07/15/open-source/fsdp/

In short, multiple parameters are combined and then split over multiple devices. The original parameter field in the module is replaced with only a view into the underlying tensor. The underlying tensor is usually empty and only re-populated (on each GPU) using the shard copies (also from each GPU), when the parameter is needed in both forward and backward computation. Therefore, there is no such thing as a fixed device for a given parameter. Each parameter are sharded to all devices conceptually.

**Comment by hyhuang00** [comment]
I see. Sorry for the confusion -- I tried to search through the documentation for how to locate a layer in a device, and the only place I got something similar is in pipeline parallelism. Thank you for your clarification -- I see the difference between them now.

I have an additional question regarding your answer. Is there a way to see the shape of the underlying tensor, and how parameters are populated into these tensors like in this figure in the blogpost?
<img width="956" alt="image" src="https://user-images.githubusercontent.com/16739711/158082128-aa086441-cc7f-4ffa-af05-ce453664a527.png">
It seems like the `LayerwiseMemoryTracker` can only produce the total memory usage during a forward/backward pass.
